src/tokenizing.py

```diff
@@ -0,0 +1,87 @@
+import grpc
+import asyncio
+
+from dataclasses import dataclass
+
+from concurrent import futures
+from proto_built_result import tokenizer_definition_pb2_grpc as tokenizer_grpc
+from proto_built_result import tokenizer_definition_pb2 as tokenizer
+
+from kiwipiepy import Kiwi
+
+kiwi_instance = Kiwi()
+
+
+@dataclass
+class TokensAndWords:
+    """
+    defining the result of tokenizing
+    """
+    words: list[str]
+    tokens: list[str]
+
+
+async def tokenize(title: str) -> TokensAndWords:
+    """
+    - handling the tokenizing process asynchronous way
+      using ThreadPool and event loop
+    - cf. the tokenizer library(Kiwi) is not supports asyncio itself
+    """
+
+    loop = asyncio.get_event_loop()
+
+    with futures.ThreadPoolExecutor() as pool:
+        parsed_tokens = await loop.run_in_executor(pool, kiwi_instance.tokenize, title.title)
+
+    tokens: list[str] = []
+    for p in parsed_tokens:
+        tokens.append(p.form)
+
+    words = title.split()
+
+    return TokensAndWords(tokens, words)
+
+
+class TokenizerServicer(tokenizer_grpc.TokenizerServicer):
+    """
+    - protobuf service layer code
+    - this service class TokenizerServicer extends TokenizerServicer
+      which generated by protoc compiler using .proto file
+    """
+
+    async def Tokenize(self, request: tokenizer.TokenizingRequest, unused_context) -> tokenizer.TokenizingResult:
+        tokenized_result = await tokenize(request.title)
+
+        protobuf_result = tokenizer.TokenizingResult()
+
+        protobuf_result.words = tokenized_result.words
+        protobuf_result.tokens = tokenized_result.tokens
+
+        return protobuf_result
+
+
+async def serve() -> None:
+    """
+    - protobuf server initiator code
+    - first, initiate grpc.aio.server(asyncio)
+    - then, add servicer to server
+    - finally, open up the port and running the server
+    """
+    server = grpc.aio.server(futures.ThreadPoolExecutor())
```

Comment:
아 여기서 저수준 API를 사용해서 전체적으로 저수준 API를 사용하셨군요. asyncio와 concurrent.futures를 섞어 쓰는 건 GIL 때문에 성능 향상이 없고 오히려 오버헤드가 발생하여 성능을 낮출 우려가 있다고 합니다. https://stackoverflow.com/questions/69728812/asyncio-and-concurrent-futures-threadpoolexecutor

Reply:
https://www.reddit.com/r/Python/comments/m7sf7y/asyncio_and_concurrentfutures/
이 부분의 2번째 답변을 보면

concurrent.futures offers a convenient API for doing work with either threads or processes. I don't really use asyncio, but I understand the equivalent is asyncio.Future.

I expect that the reason it wasn't added to the concurrent library is because of the await keyword. You can't simply mix code that uses await with code that doesn't. That's why so many libraries are being completely rewritten to work with asyncio.

To answer your main question, oversimplified a bit:

If you are doing a lot of heavy processing work, use multiprocessing (or concurrent.futures.ProcessPoolExecutor).

If you are doing occasional IO, or background work for a GUI (less than 50 tasks or so), use threading (or concurrent.futures.ThreadPoolExecutor).

If you are doing heavy IO (100+ tasks), then it makes sense to use asyncio. There is controversy over asyncio so there are alternatives, but they work fundamentally the same, just with different interfaces. Some argue that you're better off with curio or gevent (my preference, does not use await).

Reply:
라고 합니다. 음... 저라면 asyncio를 쓸 것 같아요. 왜냐하면 파이썬 재단이 지원도 많이 해주고 있고, 3.11이 되면서 Go statement considered harmful 패러다임도 도입되어 있어서 아키텍처도 깔끔하게 만들 수 있거든요.

Reply:
[gRPC 의 해당 부분](https://grpc.github.io/grpc/python/grpc_asyncio.html#create-server) 에서 `migration_thread_pool` 에 들어가는 parameter type 이 `futures.ThreadPoolExecutor` 라 확인한 gRPC 예제 코드에서 아직 더 이상의 업데이트는 없는 듯 합니다.